### PR TITLE
typing(iterators): correctly typehint iterators

### LINKED ISSUE #836: Several classes in iterators.py are untyped
### Summary

There are several classes in iterators.py where their methods and attributes have no type hints defined.

### What is the feature request for?

The core library

### The Problem

As stated in #816, not typing methods and attributes in classes makes it hard to tell for static type checkers to infer their types.

### The Ideal Solution

Type all of the methods of the following classes:

- `_ChunkedAsyncIterator`
- `_MappedAsyncIterator`
- `_FilteredAsyncIterator`
- `ReactionIterator`
- `HistoryIterator`
- `AuditLogIterator`
- `GuildIterator`
- `MemberIterator`
- `create_event` of `ScheduledEventIterator`
- `create_user` of `ScheduledEventUserIterator`

It also looks like a lot of the attributes in all of the classes defined in iterators.py have no type hints.

### The Current Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by teaishealthy**
## Summary

Closes #836 

Ref #849, `git checkout master` failed and then I git reset to upstream/master and force pushed :(

## Checklist

- [ ] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by teaishealthy** [comment]
The errors for `set_permissions` are fixed by #869

**Review by EmreTech — CHANGES_REQUESTED** [review]
This pyright dependency bump seems to be giving out a lot of errors that aren't related to this PR, so you should either move the changes to a new PR or leave the task to someone else.

**Comment by ooliver1** [comment]
> This pyright dependency bump seems to be giving out a lot of errors that aren't related to this PR, so you should either move the changes to a new PR or leave the task to someone else.

I'll get onto this now.

File: nextcord/iterators.py [hunk 1]
```diff
@@ -723,18 +760,19 @@ async def _retrieve_guilds_after_strategy(self, retrieve):
 
 
 class MemberIterator(_AsyncIterator["Member"]):
-    def __init__(self, guild, limit=1000, after=None):
+    def __init__(self, guild: Guild, limit: int = 1000, after: Optional[SnowflakeTime] = None):
```
Comment (EmreTech) [inline nextcord/iterators.py:763]: `limit` supports `None`, as shown by `MemberIterator._get_retrieve`.

**Review by EmreTech — CHANGES_REQUESTED** [review]

**Review by EmreTech — APPROVED** [review]

**Review by ooliver1 — APPROVED** [review]

**Pull request merged**